File: OctoBar/OctopusService.swift
// File: OctopusService.swift
import Foundation

/// All Octopus Energy API communication. Swift actor ensures thread safety.
actor OctopusService {
    private let session: URLSession

    private static let iso8601Formatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.timeZone = TimeZone(identifier: "Europe/London")!
        return f
    }()

    private static let ratesDecoder: JSONDecoder = {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }()

    private static let plainDecoder = JSONDecoder()

    private static let dispatchDecoder: JSONDecoder = {
        let d = JSONDecoder()
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd HH:mm:ssxxx"
        f.locale = Locale(identifier: "en_US_POSIX")
        d.dateDecodingStrategy = .formatted(f)
        return d
    }()

    init(session: URLSession = .shared) {
        self.session = session
    }

    // MARK: - Public

    /// Fetches the array of unit rates covering now ± 38 h for the account's active tariff.
    func fetchRates(apiKey: String, accountNumber: String) async throws -> [UnitRate] {
        let tariffCode  = try await fetchTariffCode(apiKey: apiKey, accountNumber: accountNumber)
        let productCode = try deriveProductCode(from: tariffCode)
        return try await fetchUnitRates(apiKey: apiKey, productCode: productCode, tariffCode: tariffCode)
    }

    // MARK: - Step 1: Resolve active tariff code

    private func fetchTariffCode(apiKey: String, accountNumber: String) async throws -> String {
        let url  = URL(string: "https://api.octopus.energy/v1/accounts/\(accountNumber)/")!
        let data = try await fetch(url: url, apiKey: apiKey)
        let response = try Self.plainDecoder.decode(AccountResponse.self, from: data)
        guard let code = response.properties
            .flatMap({ $0.electricityMeterPoints })
            .flatMap({ $0.agreements })
            .first(where: { $0.validTo == nil })?
            .tariffCode
        else { throw OctoError.noActiveTariff }
        return code
    }

    // MARK: - Step 2: Derive product code

    /// "E-1R-INTELLIGENT-GO-24-10-01-A" → "INTELLIGENT-GO-24-10-01"
    /// Drops the fuel-type + register prefix and the regional suffix.
    nonisolated func deriveProductCode(from tariffCode: String) throws -> String {
        var parts = tariffCode.components(separatedBy: "-")
        guard parts.count > 3 else { throw OctoError.invalidTariffCode(tariffCode) }
        parts.removeFirst(2)    // remove "E", "1R"
        parts.removeLast()      // remove region "A"
        return parts.joined(separator: "-")
    }

    // MARK: - Step 3: Fetch unit rates

    private func fetchUnitRates(apiKey: String, productCode: String, tariffCode: String) async throws -> [UnitRate] {
        let now = Date()
        let fmt = Self.iso8601Formatter

        var comps = URLComponents(
            string: "https://api.octopus.energy/v1/products/\(productCode)" +
                    "/electricity-tariffs/\(tariffCode)/standard-unit-rates/"
        )!
        comps.queryItems = [
            URLQueryItem(name: "period_from", value: fmt.string(from: now.addingTimeInterval(-7_200))),
            URLQueryItem(name: "period_to",   value: fmt.string(from: now.addingTimeInterval(129_600)))
        ]

        let data = try await fetch(url: comps.url!, apiKey: apiKey)
        return try Self.ratesDecoder.decode(RatesResponse.self, from: data).results
    }

    // MARK: - Intelligent Go Dispatch Slots

    /// Fetches planned smart-charge dispatch slots via the Octopus GraphQL API.
    func fetchDispatchSlots(apiKey: String, accountNumber: String) async throws -> [DispatchSlot] {
        let token = try await obtainGraphQLToken(apiKey: apiKey)
        return try await fetchPlannedDispatches(token: token, accountNumber: accountNumber)
    }

    private func obtainGraphQLToken(apiKey: String) async throws -> String {
        let url = URL(string: "https://api.octopus.energy/v1/graphql/")!
        let query = """
        mutation { obtainKrakenToken(input: { APIKey: "\(apiKey)" }) { token } }
        """
        let body = try JSONSerialization.data(withJSONObject: ["query": query])

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = body

        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse,
              (200...299).contains(http.statusCode)
        else { throw OctoError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0) }

        let result = try Self.plainDecoder.decode(GraphQLTokenResponse.self, from: data)
        return result.data.obtainKrakenToken.token
    }

    private func fetchPlannedDispatches(token: String, accountNumber: String) async throws -> [DispatchSlot] {
        let url = URL(string: "https://api.octopus.energy/v1/graphql/")!
        let query = """
        query { plannedDispatches(accountNumber: "\(accountNumber)") { startDt endDt } }
        """
        let body = try JSONSerialization.data(withJSONObject: ["query": query])

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(token, forHTTPHeaderField: "Authorization")
        req.httpBody = body

        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse,
              (200...299).contains(http.statusCode)
        else { throw OctoError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0) }

        return try Self.dispatchDecoder.decode(GraphQLDispatchResponse.self, from: data).data.plannedDispatches
    }

    // MARK: - HTTP with exponential-backoff retry (3 attempts: wait 1 s, 2 s)

    private func fetch(url: URL, apiKey: String, attempt: Int = 1) async throws -> Data {
        var req = URLRequest(url: url)
        let creds = "\(apiKey):".data(using: .utf8)!.base64EncodedString()
        req.setValue("Basic \(creds)", forHTTPHeaderField: "Authorization")

        do {
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse,
                  (200...299).contains(http.statusCode)
            else {
                throw OctoError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0)
            }
            return data
        } catch {
            guard attempt < 3 else { throw error }
            // Exponential back-off: attempt 1→wait 1s, attempt 2→wait 2s
            let delay = UInt64(pow(2.0, Double(attempt - 1))) * 1_000_000_000
            try await Task.sleep(nanoseconds: delay)
            return try await fetch(url: url, apiKey: apiKey, attempt: attempt + 1)
        }
    }
}

// MARK: - Errors

enum OctoError: Error, LocalizedError {
    case noActiveTariff
    case invalidTariffCode(String)
    case httpError(Int)
    case missingCredentials

    var errorDescription: String? {
        switch self {
        case .noActiveTariff:           return "No active electricity tariff found."
        case .invalidTariffCode(let c): return "Cannot parse tariff code: \(c)"
        case .httpError(let code):      return "HTTP error \(code)."
        case .missingCredentials:       return "API key or account number not configured."
        }
    }
}

File: OctoBar-Windows/OctopusService.cs
// File: OctopusService.cs
// All Octopus Energy API communication, ported from the Swift actor OctopusService.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace OctoBar
{
    public class OctoError : Exception
    {
        public OctoError(string message) : base(message) { }

        public static OctoError NoActiveTariff =>
            new("No active electricity tariff found.");
        public static OctoError InvalidTariffCode(string code) =>
            new($"Cannot parse tariff code: {code}");
        public static OctoError HttpError(int code) =>
            new($"HTTP error {code}.");
    }

    public class OctopusService
    {
        private readonly HttpClient _client = new();

        private static readonly JsonSerializerOptions DefaultOptions = new()
        {
            PropertyNameCaseInsensitive = true
        };

        // MARK: - Public

        /// Fetches unit rates covering now ±38 h for the account's active tariff.
        public async Task<List<UnitRate>> FetchRates(string apiKey, string accountNumber)
        {
            var tariffCode = await FetchTariffCode(apiKey, accountNumber);
            var productCode = DeriveProductCode(tariffCode);
            return await FetchUnitRates(apiKey, productCode, tariffCode);
        }

        // MARK: - Step 1: Resolve active tariff code

        private async Task<string> FetchTariffCode(string apiKey, string accountNumber)
        {
            var url = $"https://api.octopus.energy/v1/accounts/{accountNumber}/";
            var json = await Fetch(url, apiKey);
            var response = JsonSerializer.Deserialize<AccountResponse>(json, DefaultOptions)
                ?? throw new Exception("Failed to parse account response");

            var code = response.Properties
                .SelectMany(p => p.ElectricityMeterPoints)
                .SelectMany(mp => mp.Agreements)
                .FirstOrDefault(a => a.ValidTo == null)?
                .TariffCode;

            return code ?? throw OctoError.NoActiveTariff;
        }

        // MARK: - Step 2: Derive product code
        // "E-1R-INTELLIGENT-GO-24-10-01-A" → "INTELLIGENT-GO-24-10-01"

        public static string DeriveProductCode(string tariffCode)
        {
            var parts = tariffCode.Split('-').ToList();
            if (parts.Count <= 3) throw OctoError.InvalidTariffCode(tariffCode);
            parts.RemoveAt(0); // remove "E"
            parts.RemoveAt(0); // remove "1R"
            parts.RemoveAt(parts.Count - 1); // remove region suffix "A"
            return string.Join("-", parts);
        }

        // MARK: - Step 3: Fetch unit rates

        private async Task<List<UnitRate>> FetchUnitRates(
            string apiKey, string productCode, string tariffCode)
        {
            var now = DateTime.UtcNow;
            var periodFrom = Uri.EscapeDataString(now.AddHours(-2).ToString("yyyy-MM-ddTHH:mm:ssZ"));
            var periodTo   = Uri.EscapeDataString(now.AddHours(36).ToString("yyyy-MM-ddTHH:mm:ssZ"));

            var url = $"https://api.octopus.energy/v1/products/{productCode}" +
                      $"/electricity-tariffs/{tariffCode}/standard-unit-rates/" +
                      $"?period_from={periodFrom}&period_to={periodTo}";

            var json = await Fetch(url, apiKey);
            var response = JsonSerializer.Deserialize<RatesResponse>(json, DefaultOptions)
                ?? throw new Exception("Failed to parse rates response");
            return response.Results;
        }

        // MARK: - Intelligent Go Dispatch Slots

        public async Task<List<DispatchSlot>> FetchDispatchSlots(
            string apiKey, string accountNumber)
        {
            var token = await ObtainGraphQLToken(apiKey);
            return await FetchPlannedDispatches(token, accountNumber);
        }

        private async Task<string> ObtainGraphQLToken(string apiKey)
        {
            var url   = "https://api.octopus.energy/v1/graphql/";
            var query = $"mutation {{ obtainKrakenToken(input: {{ APIKey: \"{apiKey}\" }}) {{ token }} }}";
            var body  = JsonSerializer.Serialize(new { query });

            var request = new HttpRequestMessage(HttpMethod.Post, url);
            request.Content = new StringContent(body, Encoding.UTF8, "application/json");

            var response = await _client.SendAsync(request);
            if (!response.IsSuccessStatusCode)
                throw OctoError.HttpError((int)response.StatusCode);

            var json   = await response.Content.ReadAsStringAsync();
            var result = JsonSerializer.Deserialize<GraphQLTokenResponse>(json, DefaultOptions)
                ?? throw new Exception("Failed to parse token response");
            return result.Data.ObtainKrakenToken.Token;
        }

        private async Task<List<DispatchSlot>> FetchPlannedDispatches(
            string token, string accountNumber)
        {
            var url   = "https://api.octopus.energy/v1/graphql/";
            var query = $"query {{ plannedDispatches(accountNumber: \"{accountNumber}\") {{ startDt endDt }} }}";
            var body  = JsonSerializer.Serialize(new { query });

            var request = new HttpRequestMessage(HttpMethod.Post, url);
            request.Content = new StringContent(body, Encoding.UTF8, "application/json");
            request.Headers.Authorization = new AuthenticationHeaderValue(token);

            var response = await _client.SendAsync(request);
            if (!response.IsSuccessStatusCode)
                throw OctoError.HttpError((int)response.StatusCode);

            var json = await response.Content.ReadAsStringAsync();

            // Dispatch slots use a custom date format: "yyyy-MM-dd HH:mm:sszzz"
            var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
            options.Converters.Add(new DispatchDateTimeConverter());

            var result = JsonSerializer.Deserialize<GraphQLDispatchResponse>(json, options)
                ?? throw new Exception("Failed to parse dispatch response");
            return result.Data.PlannedDispatches;
        }

        // MARK: - HTTP with exponential-backoff retry (3 attempts: wait 1 s, 2 s)

        private async Task<string> Fetch(string url, string apiKey, int attempt = 1)
        {
            try
            {
                var request = new HttpRequestMessage(HttpMethod.Get, url);
                var creds   = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{apiKey}:"));
                request.Headers.Authorization = new AuthenticationHeaderValue("Basic", creds);

                var response = await _client.SendAsync(request);
                if (!response.IsSuccessStatusCode)
                    throw OctoError.HttpError((int)response.StatusCode);

                return await response.Content.ReadAsStringAsync();
            }
            catch when (attempt < 3)
            {
                var delayMs = (int)Math.Pow(2, attempt - 1) * 1000;
                await Task.Delay(delayMs);
                return await Fetch(url, apiKey, attempt + 1);
            }
        }
    }

    /// Custom converter for Octopus dispatch date format: "yyyy-MM-dd HH:mm:sszzz"
    public class DispatchDateTimeConverter : JsonConverter<DateTime>
    {
        public override DateTime Read(
            ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            var str = reader.GetString() ?? "";
            if (DateTimeOffset.TryParseExact(
                    str, "yyyy-MM-dd HH:mm:sszzz",
                    System.Globalization.CultureInfo.InvariantCulture,
                    System.Globalization.DateTimeStyles.None,
                    out var dto))
                return dto.UtcDateTime;

            return DateTime.Parse(str, null,
                System.Globalization.DateTimeStyles.RoundtripKind);
        }

        public override void Write(
            Utf8JsonWriter writer, DateTime value, JsonSerializerOptions options)
            => writer.WriteStringValue(value.ToString("yyyy-MM-dd HH:mm:sszzz"));
    }
}
